### PR TITLE
add messaging

### PR DESCRIPTION
**Pull request by Kleak**

### CODE REVIEW

File: lib/src/interop/messaging_interop.dart [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+
+
+@JS()
+@anonymous
+abstract class PayloadJsImpl {
+  external String get collapse_key;
+  external String get from;
+  external NotificationJsImpl get notification;
+}
```
Comment (kevmoo) [inline lib/src/interop/messaging_interop.dart]: Please run `dartfmt -w .`

**Comment by Kleak** [comment]
Is there a way to accept Notification in the test ?

**Comment by kevmoo** [comment]
> Is there a way to accept Notification in the test ?

I'm guessing wiring it up will be tough.

As an alternative: would you mind adding an example? Then we could validate functionality manually...

**Comment by Kleak** [comment]
Will add an example for sure ;)

**Comment by kevmoo** [comment]
Thanks!

**Comment by kevmoo** [comment]
Add a place-holder for `MESSAGING_SENDER_ID` here!

https://github.com/firebase/firebase-dart/blob/master/lib/src/assets/config.json.sample

File: example/messaging/index.html [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        label {
+            width: 100%;
+        }
+        input {
+            width: 100%;
+        }
+    </style>
+</head>
+<body>
+<div class="container">
```
Comment (kevmoo) [inline example/messaging/index.html]: Add some docs here. Where do I get the token and permission?

File: example/messaging/index.html [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        </li>
+        <li>
+            <label>
+                Notification permission
+                <input disabled id="permission" />
+            </label>
+        </li>
+    </ul>
+    <div>
+        <button id="new_notification">Send notification</button>
```
Comment (kevmoo) [inline example/messaging/index.html]: Disable this button unless folks provide things above? Are both required?

File: pubspec.yaml [hunk 4]
```diff
@@ -14,6 +14,7 @@ dependencies:
   func: ">=0.1.0 <2.0.0"
   http: '^0.11.3'
   js: "^0.6.0"
+  service_worker: ^0.1.2
```
Comment (kevmoo) [inline pubspec.yaml]: This is only used in the example, right? Should be in `dev_dependencies`

File: example/messaging/index.html (hunk 3 above)
Comment (Kleak) [reply example/messaging/index.html]: in fact the two input are informal and for testing purpose.
The permission input is only to know show if the app can send notification.
And the token one is automatically populate when you do firebase.requestPermission()
I will disable the button until the permission is granted.

File: example/messaging/index.html (hunk 2 above)
Comment (Kleak) [reply example/messaging/index.html]: These field are automatically populate by Firebase.
I show them to know the state of the app.
And you need the token to create a push notification from curl/postman/etc

File: pubspec.yaml (hunk 4 above)
Comment (Kleak) [reply pubspec.yaml]: yep will move it

File: example/messaging/index.dart [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+      : tokenInput = document.querySelector("#token"),
+        newNotification = document.querySelector('#new_notification'),
+        payloadData = document.querySelector("#payload_data"),
+        permissionInput = document.querySelector("#permission");
+
+  showMessages() async {
+    await sw.register('sw.dart.js');
+    final registration = await sw.ready;
+    final messaging = fb.messaging()
+      ..usePublicVapidKey(vapidKey)
```
Comment (kevmoo) [inline example/messaging/index.dart]: Put in a check that `vapidKey != null` – and warn the user.

I'm hitting this, but it took some digging to figure out I need to set up my `VAPID_KEY` env value

File: lib/src/assets/config.json.sample [hunk 6]
```diff
@@ -6,4 +6,6 @@
   "DATABASE_URL": "TODO",
   "STORAGE_BUCKET": "TODO",
   "PROJECT_ID": "TODO",
+  "MESSAGING_SENDER_ID": "TODO",
+  "SERVER_KEY": "TODO",
```
Comment (kevmoo) [inline lib/src/assets/config.json.sample]: Add VAPID_KEY here, too!

File: example/messaging/index.dart (hunk 5 above)
Comment (Kleak) [reply example/messaging/index.dart]: I put the check in usePublicVapidKey and if you call other messaging function that need vapidKey i throw or print a warning ?

Comment (kevmoo) [reply example/messaging/index.dart]: I'm going to create a parallel pull request which changes the call to `_config[...]` to `_getConfigValue(...)` – if the value is null, it'll throw

**Pull request merged**